### PR TITLE
Add final and const support

### PR DESCRIPTION
**Pull request by Shan1024**
## Purpose
This PR adds final variables and compile-time constant support.

Compile time constants are **top-level** nodes. Only **simple literals** are supported as the RHS at the moment.

Eg - 

```ballerina
const string name = "Ballerina"; // With type.

const age = 10; // Without type.

const A = "a";
const string B = "b";

type AB A|B;

function isA(AB ab) returns boolean {
    if (ab == A) {
        return true;
    }
    return false;
}

const value = getValue(); // Invalid.
```

`final` has been added as a keyword and `@final` annotation has been removed. Only **top-level nodes**(except channels) can be declared as final.

Eg -

```ballerina
final string name = "Ballerina"; // With type.

final var age = 10; // Without type.

final string value = getValue(); // Valid.
```

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/elements/Flag.java [hunk 1]
```diff
@@ -90,5 +90,9 @@
     /**
      * Indicates Flagged node is a abstract node.
      */
-    ABSTRACT
+    ABSTRACT,
+    /**
+     * Indicates Flagged node is a compile time constant.
+     */
+    COMPILE_TIME_CONSTANT
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/model/elements/Flag.java]: Should we say Const?

File: compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 [hunk 2]
```diff
@@ -150,6 +150,7 @@ annotationDefinition
 
 globalVariableDefinition
     :   (PUBLIC)? typeName Identifier (ASSIGN expression )? SEMICOLON
+    |   (PUBLIC)? CONST typeName Identifier ASSIGN expression SEMICOLON
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4]: typename can be optional here.

File: compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 [hunk 3]
```diff
@@ -304,6 +305,7 @@ statement
 
 variableDefinitionStatement
     :   typeName Identifier (ASSIGN expression)? SEMICOLON
+    |   CONST typeName Identifier ASSIGN expression SEMICOLON
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4]: I think we don't need const in statements.

**Comment by Shan1024** [comment]
Closing the PR since some changes are planned.

**Pull request closed without merging**